### PR TITLE
Add concept of StrictMode levels 1 & 2 (no public API yet)

### PR DESCRIPTION
**Pull request by bvaughn**
### PR Stack
- [x] **#20844: Add concept of StrictMode levels 1 & 2 (no public API yet)**
- [x] #20847: Separate strict effects mode feature flags
- [x] #20849: Add StrictMode level prop and createRoot unstable_strictModeLevel option

---

## Strict mode level 0 (aka loose or sloppy mode)
This isn't really a mode so much as it's the absence of any strict modes.

## Strict mode level 1 (aka legacy mode)
The exported `<React.StrictMode>` tag remains the same and opts legacy subtrees into strict mode level one (`mode == StrictLegacyMode`). This mode enables DEV-only double rendering, double component lifecycles, string ref warnings, legacy context warnings, etc. The primary purpose of this mode is to help detected render phase side effects. No new behavior. Roots created with experimental `createRoot` and `createBlockingRoot` APIs will also (for now) continue to default to strict mode level 1.

## Strict mode level 2 (aka strict effects mode)
In a subsequent commit I will add support for a `"level"` attribute on the `<React.StrictMode>` tag (as well as a new option supported by `createRoot`). This will be the way to opt into strict mode level 2 (`mode == StrictEffectsMode`). This mode will enable DEV-only double invoking of effects on initial mount (see #19523). This will simulate future Offscreen API semantics for trees being mounted, then hidden, and then shown again. The primary purpose of this mode is to enable applications to prepare for compatibility with the new Offscreen API (more information to follow shortly).

## Strict mode level >= 3?
In the future we may add additional strict mode levels (TBD).

## External changes?
For now, this commit changes no public facing behavior. The only mechanism for opting into strict mode level 2 is the pre-existing `enableDoubleInvokingEffects` feature flag (only enabled within Facebook for now).

### CODE REVIEW

File: packages/react-reconciler/src/ReactTypeOfMode.js [hunk 1]
```diff
@@ -9,11 +9,11 @@
...
-export const ProfileMode = 0b01000;
-export const DebugTracingMode = 0b10000;
+export const NoMode = /*           */ 0b000000;
+// TODO: Remove BlockingMode and ConcurrentMode by reading from the root tag instead
+export const BlockingMode = /*     */ 0b000001;
+export const ConcurrentMode = /*   */ 0b000010;
+export const ProfileMode = /*      */ 0b000100;
+export const DebugTracingMode = /* */ 0b001000;
+export const StrictModeL1 = /*     */ 0b010000;
+export const StrictModeL2 = /*     */ 0b100000;
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactTypeOfMode.js:19]: Another way to do this is to keep each category (render phase, effects, etc) as a separate bit

but then in the <StrictMode /> implementation, assign the correct bits for each level:

```js
if (level >= 1) {
  mode |= StrictRender;
}
if (level >= 2) {
  mode |= StrictEffects;
}
```

That way there's a single place where can decide which category of warnings goes into which level.

Comment (bvaughn) [reply packages/react-reconciler/src/ReactTypeOfMode.js:19]: Yeah, considered something along those lines but it seemed like it would balloon into a lot of flags (at least for level 1).

Are you just sharing this as an idea for consideration or do you think it's objectively better?

Comment (acdlite) [reply packages/react-reconciler/src/ReactTypeOfMode.js:19]: I think effectively we're going to have one mode bit per level anyway, at least for what we have planned (just strict render and strict effects, right). So this does feel better to me.

Re: ballooning, yeah at some point we'll have to deal with running out of bits. But we have 32 so that will probably take a while.

Comment (bvaughn) [reply packages/react-reconciler/src/ReactTypeOfMode.js:19]: Maybe put another way: Seems unlikely that we're going to go back and _change_ what base strict mode means, so not much value in extracting the legacy strict mode things. Maybe future things (l2+)?

Comment (bvaughn) [reply packages/react-reconciler/src/ReactTypeOfMode.js:19]: So that case... maybe `StrictRender` becomes `StrictLegacy` and L2 is `StrictEffects` and L3 is `StrictWhateverElse`?

Comment (bvaughn) [reply packages/react-reconciler/src/ReactTypeOfMode.js:19]: > I think effectively we're going to have one mode bit per level anyway, at least for what we have planned (just strict render and strict effects, right). So this does feel better to me.

Ah. I think I misunderstood your suggestion for L1 maybe. I thought you were suggesting:
```js
if (levelProp >= 1) {
  mode |= StrictRender;
  mode |= DoubleLifecycles;
  mode |= NoLegacyContext;
  // etc.
}
if (levelProp >= 2) {
  mode |= StrictEffects;
}
```

Seems like you're more just suggesting a mode-name change?

"StrictRender" is a little weird still since it controls more than just that. That's kind of why I like `StrictLegacy` or something.

Comment (acdlite) [reply packages/react-reconciler/src/ReactTypeOfMode.js:19]: > "StrictRender" is a little weird still since it controls more than just that.

My thinking was that it's designed to help discover render phase mutations/effects.

But yeah it's fine to group them like that. My other point was just that level prop of 2 turns on both mode bits.

Comment (bvaughn) [reply packages/react-reconciler/src/ReactTypeOfMode.js:19]: > But yeah it's fine to group them like that. My other point was just that level prop of 2 turns on both mode bits.

Oh yeah, for sure this is the case regardless of naming. 👍 

Was just planning on adding that new prop in a follow up PR.

Comment (bvaughn) [reply packages/react-reconciler/src/ReactTypeOfMode.js:19]: Okay. Renamed like we talked about.

I'll do the level prop follow up next.

**Comment by dai-shi** [comment]
StrictMode level sounds nice.
I hope it would be possible to add useRef read warning on mutable values, at some level.
https://github.com/facebook/react/pull/18545#issuecomment-619264421

**Comment by bvaughn** [comment]
IMO ref mutation warning should be outside of strict mode. (But I guess you could argue it belongs in level 1.)

Mutable Source mutation belongs in level 1 too.

**Comment by jacobp100** [comment]
Sorry if this is too off-topic - does this mean double invoking the functions passed to `useEffect`?

**Comment by eps1lon** [comment]
> Sorry if this is too off-topic - does this mean double invoking the functions passed to `useEffect`?

@jacobp100 For what "strict effects mode" is, check out the original PR: 

TL;DR: It doesn't just double invoke the functions passed to `useEffect` (`create`) but also runs their cleanup logic inbetween:

> This PR changes the behavior in DEV when an effect runs from `create()` to `create() -> destroy() -> create()`

--https://github.com/facebook/react/pull/19523

**Comment by bvaughn** [comment]
Circle CI doing something weird here I think. Hm. Should sort itself out next time I rebase onto master.

**Review by rickhanlonii — COMMENTED** [review]
I still wonder if we can wait on this until we have more information but happy to move forward if others want to approve and land.

File: packages/react-reconciler/src/ReactChildFiber.new.js [hunk 2]
```diff
@@ -46,7 +46,7 @@ import {
 } from './ReactFiber.new';
 import {emptyRefsObject} from './ReactFiberClassComponent.new';
 import {isCompatibleFamilyForHotReloading} from './ReactFiberHotReloading.new';
-import {StrictMode} from './ReactTypeOfMode';
+import {StrictLegacyMode} from './ReactTypeOfMode';
```
Comment (rickhanlonii) [inline packages/react-reconciler/src/ReactChildFiber.new.js:49]: Chatted offline and we can follow up with this if we agree - we may want to name this something other than "legacy" since the term overloaded to mean "legacy root", "legacy strict mode" and any other "legacys" we have. I was thinking something like:

- `NoMode` = 0
- `StrictWarningMode` = Level 1
- `StrictRenderMode` = Level 2
- `StrictEffectMode` = Level 3

**Review by rickhanlonii — APPROVED** [review]
Looks great, these changes are really exciting 🎉

**Comment by bvaughn** [comment]
Merged via #20849
